style: Fix raise-without-from-inside-except (B904)

echoix opened the pull request:
This PR I want a little bit more attentive review.


Ruff rule: https://docs.astral.sh/ruff/rules/raise-without-from-inside-except/
The flake8 bugbear README also refers to the exception chaining tutorial here: https://docs.python.org/3/tutorial/errors.html#exception-chaining, quite useful to understand what exception chaining is (and is a very good thing for an application like ours when exceptions are rethrown to have more useful details for a user instead of scaring them off).


I wasn't completely sure to understand the existing patterns of exceptions, especially on how they were used.
For the ImportError one, there is no problem, it's obvious that it's good now.
Places where we wouldn't want exception chaining, and also not having the "During handling of the above exception, another exception occurred:" message when raising another exception in an except, we can use the "from None" idiom, and only the exception raised (inside the except) will be shown. Like in the python docs:


> It also allows disabling automatic exception chaining using the from None idiom:
> 
> ```pycon
> >>> try:
> ...    open('database.sqlite')
> ... except OSError:
> ...     raise RuntimeError from None
> 
> Traceback (most recent call last):
>   File "<stdin>", line 4, in <module>
> RuntimeError
> ```
> _https://docs.python.org/3/tutorial/errors.html#exception-chaining_





---

## Updated scope (by @wenzeslaus)

The branch has been merged up to current `main` (not rebased, to keep the shared
history intact). The non-style changes that were originally part of this PR were
split into separate PRs so they could be reviewed and merged on their own; all
four are now merged:

- #7688 (wxGUI): the matplotlib import hints, now added with `exception.add_note()`
- #7689 (grass.gunittest): the test loader import error, now a note (also drops a
  Python 2 `e.message` leftover)
- #7690 (grass.pygrass): the psycopg2 import hint, now a note
- #7687 (grass.utils): the download error message, URL scheme allowlist, and tests

Those merged versions supersede the corresponding hunks that were here. This PR
enables the `B904` rule and adds the `from ...` chaining across the rest of the
codebase. Two hunks in files the split PRs only partly touched remain here:
`from error` on `extract_zip`'s `BadZipfile` in `download.py`, and `from exc` in
`table.py` `execute()`.

Exceptions are chained with `from error` to preserve the original, including in
the places where the original is also passed into the message of the new
exception, so the detail appears both in the message and in the chained
traceback; reducing that duplication can be done later. `from None` is used where
the original would be noise: where it is an implementation detail of the raising
code (for example, `IndexError` becoming `StopIteration`) or where the new
exception gives a clearer replacement message (for example, an invalid value for
a database column). In wxGUI, the original error is passed into the new exception
and `from None` is used, because that exception is caught immediately and only its
message is shown, so no traceback is produced.

The three behavior fixes from the original PR are retained: the rdigit
`ScriptError` fix, the removed unreachable `return` in the toolbars, and the
removed dead `try`/`except` in `__ll_parts` (see the commit message).


## Commit message

Suggested squash commit message:

```
style: Fix raise-without-from-inside-except (B904)

Enable the ruff B904 rule and add explicit exception chaining to raise
statements inside except clauses.

Most exceptions are chained with "from error" to preserve the original,
including where the original is also passed into the message of the new
exception, so the detail is both in the message and in the chained traceback;
reducing that duplication is left for later. "from None" is used where the
original would be noise: where it is an implementation detail of the raising
code (for example, IndexError becoming StopIteration) or where the new
exception gives a clearer replacement message (for example, an invalid value
for a database column). In wxGUI, the original error is passed into the new
exception and "from None" is used; that exception is caught immediately and
only its message is shown, so no traceback is produced.

The change covers wxGUI, grass.script, grass.pygrass, grass.gunittest,
grass.imaging, grass.pydispatch, grass.semantic_label, grass.utils, and
r.in.wms.

Related bug fixes included:

* rdigit: "raise ScriptError" raised the class, not an instance. ScriptError
  requires a message argument, so this raised TypeError instead, which the
  callers' "except ScriptError" handlers did not catch. A failing r.mapcalc
  or g.copy resulted in a traceback instead of showing an error message. The
  message of the module error is now passed to ScriptError.
* wxGUI toolbars: Remove an unreachable return statement after a raise.
* wxGUI utils: Remove a dead try-except in __ll_parts which could not be
  reached and only re-raised an empty ValueError.
```

echoix committed:
style: Fix raise-without-from-inside-except (B904)

echoix commented:
@ninsbl That's it for now, I don't have any more queued changes from the last week/weekends that are ready to be converted to a PR. I have one that needs to be applied after the pydispatch code synced from upstream, and another one that is maybe incomplete (it isn't done by ruff, it is manual from the actual pylint failures as ruff doesn't have that rule implemented yet, and since I never got to the point that the first pylint-related step in CI passes, I don't see if there's more.
My third set of changes are still a work-in-progress, as it's a bit a bigger change to make the (ruff) suggestion work in our context. (Since it removes some derived classe's __init__, I have to adjust the docstrings around and check that the sphinx-generated output is still fine. For now it isn't to my liking.

So, it'll take a couple days to have some more changes ready to review.

ninsbl commented on gui/wxpython/tplot/frame.py:
```diff
@@ -46,7 +46,7 @@
             "(python-matplotlib) package to be installed. {0}"
         ).format(e)
-    )
+    ) from e
```
I tried to understand the issue. Not sure if I got it completly right. My interpretation would be, that authors did not want excepting-chaining when they included the previous exception text with `format` in the raised exception...
So `from e` would print _e_ twice in this case, no? So probably, `from None` is the right choice when the original Error message is included in the raised exception? Not sure when one should raise a python error instead of a eg. _gs.fatal(_())_ ... Maybe @wenzeslaus has some thoughts on exception chaining?

echoix replied:
I don't think exception chaining was used at all before. We're way more precise about our python usage now than when the python code started, so I'm pretty sure not everything was thoroughly though out before.

But for this case, maybe forgetting the `.format(e)` and using exception chaining for the details is in the same spririt?
The spirit (as I understand), was to have better error messages that what python was giving at the time. Now, especially in the latests versions, have been giving better error messages, so it's less frightening and more useful. I think there's value of having the real errors showed up (as we will be having that in the logs in the issues the users will file), and we might want to have all that extra context.

echoix replied:
Oh, it's only a hint for an import error. So that's kinda easy and has no real implications for the logic of the code itself. It's only a nice-to-have hint.

So, I'm suggesting exception chaining without the format(e) here, it will give something similar to the example of python docs:
```pycon
>>> def func():
...    raise ConnectionError
... 
>>> try:
...     func()
... except ConnectionError as exc:
...     raise RuntimeError('Failed to open database') from exc
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "<stdin>", line 2, in func
ConnectionError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
RuntimeError: Failed to open database
``` 

```pycon
>>> def func():
...    raise ConnectionError
... 
>>> try:
...     func()
... except ConnectionError as exc:
...     raise RuntimeError('Failed to open database') from exc
... 
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "<stdin>", line 2, in func
ConnectionError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
RuntimeError: Failed to open database
```


```python
try:
    import matplotlib as mpl
    # The recommended way to use wx with mpl is with the WXAgg
    # backend.
    mpl.use("WXAgg")
    from matplotlib.figure import Figure
    from matplotlib.backends.backend_wxagg import (
        FigureCanvasWxAgg as FigCanvas,
        NavigationToolbar2WxAgg as NavigationToolbar,
    )
    import matplotlib.dates as mdates
except ImportError as e:
    raise ImportError(
        _(
            'The Temporal Plot Tool needs the "matplotlib" '
            "(python-matplotlib) package to be installed."
        )
    ) from e
```

giving (except for the line numbers and file names, I wrote it by hand):
```pycon
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "<stdin>", line 2, in func
ImportError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
ImportError: The Temporal Plot Tool needs the "matplotlib" (python-matplotlib) package to be installed.
```

echoix replied:
Or maybe we'd want to use exception notes: https://docs.python.org/3/tutorial/errors.html#enriching-exceptions-with-notes?

ninsbl replied:
Both exception chaining and exception notes seem OK to me. However, I tend to think that all tracebacks also could be considered bugs, and when we catch an exception with try except, a simple (error-) message should be given that is addressed to the user and not to a developer who should be able to read traceback...
Maybe this discussion would make a good startingpoint to improve the exception-handling section in the style-guide?!?

echoix replied:
For import errors, it seems weird to me that we use dependencies, but don't list them or require them as any other python package. But that's a bit because we don't have a real python package neither, and it brings these kinds of problems.

echoix replied:
We shouldn't be trying to have nicer errors for import errors, if the requirements should've been installed normally and automatically as a dependency). It isn't the case right now, so it's fine to have these.

I think I'll add a TODO in the changed code. But I'll use exception chaining to align with the rest of the PR. Exception notes might be a good idea for all other cases otherwhere, but not in this.

ninsbl replied:
As for dependencies, see also https://trac.osgeo.org/grass/ticket/2895

echoix replied:
Even then, that ticket is starting to get old, now everything should be in a pyproject.toml. For addons, I think each should be an individual "real" python package, each with a pyproject.toml, and could be fetched only for that subdirectory and work perfectly (and allow code tools to understand that structure).

ninsbl replied:
Making addons python packages is also the way QGIS is going, I think. However, we also have C-addons and addons using R...

echoix replied:
I didn't go with the TODO finally. Still working on the next PR, the unsafe fixes are a little longer to go through, and I have to do more manual tuning.

ninsbl replied:
@neteler do you have any opions/thoughts on raised exceptions, exception chaining vs. GRASS messages (fatal/error)?
See also my comment here: https://github.com/OSGeo/grass/pull/4032#discussion_r1676829971

wenzeslaus replied:
We need to do the discussion of exception handling separately, ideally together with style guide updates as @ninsbl suggests.

wenzeslaus replied:
Generally, library should raise exceptions. Tools should show error message. Tools don't need to account for programming errors by showing error messages, i.e., for programing error, traceback is completely appropriate.

echoix committed:
Update frame.py

ninsbl commented on python/grass/script/task.py:
```diff
@@ -502,7 +502,7 @@ def get_interface_description(cmd):
                 "\n\nDetails: <{det}>"
             ).format(cmd=cmd, det=e)
-        )
+        ) from e
```
Here is another (of several) examples, where the original Error messages is included in the error message that is raised at the end. So, I guess we also here could replace the formating with exception chaining?

echoix replied:
Maybe this one with a `from None`?

wenzeslaus replied:
The motivation for these `Details:` is to give user the original exception. This would be better done with chained exceptions in Python. However, the code was written with the expectation that the error will be read by a command line tool user without a traceback (or the original exception). Using `from None` preserves that idea, but I would be open to a solution which goes the path with less expectations about the potential usage.

Perhaps `from e` combined with the `Details:` has most information when doing `print(...)` and when just looking at the traceback.

wenzeslaus replied:
I did several PRs already to address different edge cases here, and this would be yet another deep rabbit hole, so let's leave this as is. Per my previous comment `from None` would make more sense with dropping the `Details: <error>` but that would loose info in a simple print which will not print the traceback. This would be for a more broad review of exception propagation and printing. Leave `from e` which is duplicative, but does not loose any info anywhere and helps debugging by the stacked tracebacks.

ninsbl commented on python/grass/script/core.py:
```diff
@@ -1939,7 +1939,7 @@ def _create_location_xy(database, location):
         os.chdir(cur_dir)
     except OSError as e:
-        raise ScriptError(repr(e))
+        raise ScriptError(repr(e)) from e
```
This basically looks like a pointless try except, as the original Error is raised, so why using try-except here, if not the scripterror simplifies the error message presented to the user...

wenzeslaus replied:
ScriptError seems to be the error which is raised by related functions, so I think is is for consistency.

With that said, OSError seems like a fine exception to be raised from project creation functions (which would remove the need for try-except).

wenzeslaus replied:
Let's keep this for consistency. The _set_location_description function does the same with OSError. Replacing it would be a larger, separate effort.

ninsbl reviewed (commented):
I think we touch upon some general questions here, where I would love to hear the opinion of e.g. @wenzeslaus or @petrasovaa ...

ninsbl commented on python/grass/pygrass/modules/interface/module.py:
```diff
@@ -555,7 +555,7 @@ def __init__(self, cmd, *args, **kargs):
             print("OSError error({0}): {1}".format(e.errno, e.strerror))
             str_err = "Error running: `%s --interface-description`."
-            raise GrassError(str_err % self.name)
+            raise GrassError(str_err % self.name) from e
```
```suggestion
            raise GrassError(str_err % self.name) from None
```
Maybe here, and in similar cases too?

echoix replied:
The code around here is weird, we shouldn't be using print for this there. (It was out of scope of the PR, so I wasn't changing everything, in order to be reviewable). But None makes sense, as OSError is only to know what that we couldn't find the module (or fails), it's not the real exception.

wenzeslaus replied:
You can get PermissionError or FileNotFoundError which is likely a relevant info. The `from e` will preserve that, no? ...if yes, that would be an improvement. Semantic of `GrassError` is not clear to me and `Error running:` is more hiding than explaining what happened here. The `print` seems to assume some usage and tries to add convenience. ...yep, some deeper revision would be nice.

ninsbl commented on python/grass/pygrass/rpc/base.py:
```diff
@@ -173,7 +173,9 @@ def safe_receive(self, message):
-            raise FatalError("Exception raised: " + str(e) + " Message: " + message)
+            raise FatalError(
+                "Exception raised: " + str(e) + " Message: " + message
+            ) from e
```
```suggestion
            ) from None
```

echoix replied:
This one I'm not sure we'd want this, because there are multiple types of errors caught. So the traceback will be able to tell us which one is the direct cause of the others

wenzeslaus replied:
Similar case as the other one with `Details: <error>`. We need one or the other, but which one depends on proper handling, propagation, and consumption (printing) of these exceptions. `from e` creates duplication, but it seems correct for current Python, while `Exception raised: <error>` is what is up for later revision.

echoix commented:
If we are to address each of these for the best solution in context, it would make sense to expand the scope of the PR and use exception notes where it makes more sense.

echoix committed:
Merge branch 'main' into fix-raise-without-from-inside-except

wenzeslaus commented:
## Raising exceptions versus printing error messages

Printing error messages is for command line tools. Python functions should generally raise exceptions.

The Python code in tools (modules; both in core and addons) should generally print an error message calling `fatal`. User will get the error in command line, GUI, or stderr in a subprocess call in Python. The `fatal` is a shortcut how to print an error message in a standardized way and end the execution of the process. A command line tool has a pretty good idea when to end the process.

Python functions in the library should raise exceptions to give the caller freedom to react to them in whatever way is appropriate. This is what is expected from a Python library as exceptions are the standard way in Python of reporting errors and a single function in a library does not have enough context to do anything more specific.

Tracebacks may be perceived by users as errors and are less readable then plain error messages, so it is desirable for the command line tool to catch exceptions and print error messages instead. This may also allow for switch from Python and function oriented wording to one more appropriate for a command line tool. However, not all possible exceptions need to be caught. Unlikely and unexpected exceptions unrelated to the standard execution of the program, e.g., programmer's errors, are best left uncaught so that full information in a debugging-friendly format, i.e., traceback, is displayed. The situation in the GUI is similar, although the dividing line is harder to spot and the details would be for a separate discussion.

This library-tool distinction gets complicated with how the library functions in the `grass` package were envisioned. Similarly to C, the idea for some of these functions was that they are solely meant for direct support of development of command line tools (I suppose; I was not here for that). Hence, it seemed convenient to prefer calls to `fatal` over raising exceptions. Given the benefits for writing tools, the practice spread widely at least over `grass.script` which would prevent its full use as a library, but at the same time, a function was introduced to control the behavior of `fatal` which now by default calls exit, but can also be set to raise an exception allowing its full use as a library. (We can consider for v9 whether the default should be changed.)

wenzeslaus commented on python/grass/pygrass/vector/table.py:
```diff
@@ -859,10 +859,10 @@ def connection(self):
                 return psycopg2.connect(db)
             except ImportError:
                 er = "You need to install psycopg2 to connect with this table."
-                raise ImportError(er)
+                raise ImportError(er) from None
         else:
             str_err = "Driver is not supported yet, pleas use: sqlite or pg"
-            raise TypeError(str_err)
+            raise TypeError(str_err) from None
```
Wouldn't the `except ImportError: raise ImportError` cases be best handled by `add_note...raise`? I would say yes, except that we need to wait for min Python on main version to be raised to 3.11 ([PEP](https://peps.python.org/pep-0678/)).

echoix replied:
It's always possible to use the feature if it is available and let it go if not supported. Or do a fallback of some sort, that adds the text to the message instead.

wenzeslaus replied:
We have 3.11 as a min now, so `add_note...raise` was used.

wenzeslaus commented on python/grass/script/core.py:
```diff
@@ -867,8 +867,8 @@ def _parse_opts(lines):
             var, val = line.split(b"=", 1)
-        except ValueError:
-            raise SyntaxError("invalid output from g.parser: {}".format(line))
+        except ValueError as err:
+            raise SyntaxError("invalid output from g.parser: {}".format(line)) from err
```
Now the code uses `as e`, `as err` and `as error`. I suggested to use `as error` in the new code (a single, existing word).

echoix replied:
I think there were some name collisions when only one name is used, but I'm not sure.

echoix replied:
Renaming these variables to be uniform would be something to do throughout the code.

wenzeslaus replied:
Hm, right, maybe not this PR.

echoix committed:
Merge branch 'main' into fix-raise-without-from-inside-except

echoix committed:
python: Fix raise-without-from-inside-except (B904)

Rule contents: Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

echoix committed:
checks: Enable checking for ruff rule B904

echoix committed:
Merge branch 'main' into fix-raise-without-from-inside-except

echoix committed:
Merge branch 'main' into fix-raise-without-from-inside-except

echoix committed:
Merge branch 'main' into fix-raise-without-from-inside-except

echoix committed:
Merge branch 'main' into fix-raise-without-from-inside-except

wenzeslaus committed:
Merge remote-tracking branch 'upstream/main' into fix-raise-without-from-inside-except

wenzeslaus committed:
wxGUI: Remove useless try-except in __ll_parts

The last fallback branch only indexes and assigns strings, so it cannot
raise ValueError and the handler was dead code re-raising an empty
ValueError.

wenzeslaus committed:
grass.utils: Use one name for the caught exception in download_and_extract

The two adjacent handlers used 'err' and 'e' for the same kind of error.
The rest of the file and g.extension use 'error'.

wenzeslaus committed:
grass.pygrass: Do not suppress exception context outside of an except clause

The raise is in the function body, not in an except clause, so B904 does
not apply to it. The 'from None' there would hide the original exception
if connection() is called while another exception is being handled.

wenzeslaus committed:
grass.script: Chain AttributeError in KeyValue.__getattr__

This raise appeared on main after this branch was created and is the only
remaining B904 violation, so the rule cannot be enabled without it. The
KeyError is an implementation detail of the lookup, so it is suppressed.

wenzeslaus commented:
I'm preparing an review-update here to finish things up, but it will depend on #7684.

wenzeslaus commented:
Now this depends on #7688, #7689 and #7690.

wenzeslaus committed:
Merge remote-tracking branch 'upstream/main' into HEAD

wenzeslaus committed:
Merge remote-tracking branch 'upstream/main' into fix-raise-without-from-inside-except

# Conflicts:
#	gui/wxpython/iscatt/plots.py
#	gui/wxpython/timeline/frame.py
#	gui/wxpython/tplot/frame.py
#	python/grass/gunittest/loader.py
#	python/grass/pygrass/vector/table.py
#	python/grass/utils/download.py

wenzeslaus committed:
Merge remote-tracking branch 'upstream/main' into fix-raise-without-from-inside-except

wenzeslaus commented:
There was a lot of related PRs to finalize this (in addition to the commits I made):

Split out of this one (merged):
- #7687 grass.utils: download error message and URL schemes
- #7688 wxGUI: matplotlib hint as an exception note
- #7689 grass.gunittest: module name as an exception note
- #7690 grass.pygrass: psycopg2 hint as an exception note

Minimal Python version (the exception notes above need Python 3.11):
- #7684 packaging: update the minimal Python version to 3.11
- #7691 doc: update RFC 8 on the minimal Python version in addons (the last one open at the time of writing needs PSC approval for merge)
- OSGeo/grass-addons#1785 packaging: explain the minimal Python version

echoix reviewed (commented):
@wenzeslaus I can't approve my own PR, so this is an approval of what's remaining
